prov/gni/include/gnix_vector.h

```diff
@@ -0,0 +1,456 @@
+				      gnix_vec_index_t index)
+{
+	GNIX_TRACE(FI_LOG_EP_CTRL, "\n");
+
+	if (unlikely(!vec || vec->state == GNIX_VEC_STATE_DEAD || !entry ||
+		    index >= vec->attr.vec_maximum_size)) {
+		GNIX_WARN(FI_LOG_EP_CTRL, "Invalid parameter to "
+			  "__gnix_vec_insert_at\n");
+		return -FI_EINVAL;
+	} else {
+		if (index >= vec->attr.cur_size) {
+			uint64_t new_size = __gnix_vec_get_new_size(vec, index);
+			int ret = __gnix_vec_resize(vec, new_size);
+
+			if (unlikely(ret))
+				return ret;
+		}
+
+		if (vec->vector[index]) {
+			GNIX_WARN(FI_LOG_EP_CTRL, "Existing element found in "
+				  "__gnix_vec_insert_last\n");
+			return -FI_ECANCELED;
+		} else {
+			vec->vector[index] = entry;
+			return FI_SUCCESS;
+		}
+	}
+}
+
+/**
+ * Removes the first element from the vector and destroys it.  Note that
+ * this function destroys the element even if that element is still connected
+ * or in use.
+ *
+ * @param[in] vec	the vector to remove an entry from
+ *
+ * @return FI_SUCCESS	 Upon successfully removing and destroying the entry
+ * @return -FI_EINVAL	 Upon passing a dead entry
+ * @return -FI_ECANCELED Upon attempting to remove an empty entry
+ */
+static inline int _gnix_vec_remove_first(gnix_vector_t *vec)
+{
+	return _gnix_vec_remove_at(vec, 0);
+}
+
+/**
+ * Removes the last element from the vector and destroys it.  Note that
+ * this function destroys the element even if that element is still connected
+ * or in use.
+ *
+ * @param[in] vec	the vector to remove an entry from
+ *
+ * @return FI_SUCCESS	 Upon successfully removing and destroying the entry
+ * @return -FI_EINVAL	 Upon passing a dead entry
+ * @return -FI_ECANCELED Upon attempting to remove an empty entry
+ */
+static inline int _gnix_vec_remove_last(gnix_vector_t *vec)
+{
+	return _gnix_vec_remove_at(vec, vec->attr.cur_size-1);
+}
+
+/**
+ * Removes the element at index from the vector and destroys it.  Note that
+ * this function destroys the element even if that element is still connected
+ * or in use.
+ *
+ * @param[in] vec	the vector to remove an entry from
+ * @param[in] index	the index of the entry to remove
+ *
+ * @return FI_SUCCESS	 Upon successfully removing and destroying the entry
+ * @return -FI_EINVAL	 Upon passing a dead vector
+ * @return -FI_ECANCELED Upon attempting to remove an empty entry or failing to
+ *			 properly destroy a vector element.
+ */
+static inline int _gnix_vec_remove_at(gnix_vector_t *vec, gnix_vec_index_t index)
+{
+	int ret;
+
+	GNIX_TRACE(FI_LOG_EP_CTRL, "\n");
+
+	if (unlikely(!vec || vec->state == GNIX_VEC_STATE_DEAD || index >= vec->attr.cur_size)) {
+		GNIX_WARN(FI_LOG_EP_CTRL, "Invalid parameter to "
+			  "__gnix_vec_remove_at\n");
+		return -FI_EINVAL;
+	} else {
+		if (!vec->vector[index]) {
+			GNIX_WARN(FI_LOG_EP_CTRL, "No entry exists in "
+				  "__gnix_vec_remove_at\n");
+			return -FI_ECANCELED;
+		} else {
+			ret = vec->attr.destructor(vec->vector[index]);
+			if (ret) {
```

Comment:
For convenience.  Although, I think this makes the vector more confusing to use unless we add a constructor as well.  I believe the constructor pointer would be declared something like `int (*constructor)(...)`.  However, I'm not sure how to call the constructor within the vector without adding a variable argument list to the `insert` functions.  Any suggestions?

On Fri, 04 Mar 2016 08:19:46 -0800
James Swaro <notifications@github.com> wrote:

> > +	int ret;
> > +
> > +	GNIX_TRACE(FI_LOG_EP_CTRL, "\n");
> > +
> > +	if (unlikely(!vec || vec->state == GNIX_VEC_STATE_DEAD || index >= vec->attr.cur_size)) {
> > +		GNIX_WARN(FI_LOG_EP_CTRL, "Invalid parameter to "
> > +			  "__gnix_vec_remove_at\n");
> > +		return -FI_EINVAL;
> > +	} else {
> > +		if (!vec->vector[index]) {
> > +			GNIX_WARN(FI_LOG_EP_CTRL, "No entry exists in "
> > +				  "__gnix_vec_remove_at\n");
> > +			return -FI_ECANCELED;
> > +		} else {
> > +			ret = vec->attr.destructor(vec->vector[index]);
> > +			if (ret) {
> 
> Why is the vector responsible for destroying an item when the item was created outside of the context of the vector?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/ofi-cray/libfabric-cray/pull/696/files#r55050455


-- 
Evan Harvey